core/src/main/scala/cats/instances/hash.scala

```diff
@@ -0,0 +1,23 @@
+package cats
+package instances
+
+import cats.kernel.instances._
+
+trait HashInstances {
+  implicit val catsContravariantCartesianForHash: ContravariantCartesian[Hash] =
+    new ContravariantCartesian[Hash] {
+      /**
+       * Derive an `Hash` for `B` given an `Hash[A]` and a function `B => A`.
+       */
+      def contramap[A, B](ha: Hash[A])(f: B => A): Hash[B] = Hash.by(f)(ha)
+
+      /**
+       * Derive an `Hash` for `(A, B)` given `Hash[A]` and `Hash[B]`.
+       */
+      def product[A, B](ha: Hash[A], hb: Hash[B]): Hash[(A, B)] = new Hash[(A, B)] {
```

Comment:
untested

Reply:
this is still untested.

Can we test the cartesian laws?

Also, can we test that `product(h1, h2).hash((a, b)) == implicitly[Hash[(A, B)]].hash((a, b))` I think it would be weird if there are two different ways to compose hashes.